[AMDGPU] introduce S_WAITCNT_LDS_DIRECT in the memory legalizer

ssahasra opened the pull request:
The new instruction represents the unknown number of waitcnts needed at a
release operation to ensure that prior direct loads to LDS (formerly called LDS
DMA) are completed. The instruction is replaced in SIInsertWaitcnts with a
suitable value for vmcnt().

Co-authored-by: Austin Kerbow <austin.kerbow@amd.com>.

ssahasra committed:
[AMDGCN] pre-checkin test for LDS DMA and release operations

ssahasra committed:
[AMDGPU] introduce S_WAITCNT_LDS_DIRECT in the memory legalizer

The new instruction represents the unknown number of waitcnts needed at a
release operation to ensure that prior direct loads to LDS (formerly called LDS
DMA) are completed. The instruction is replaced in SIInsertWaitcnts with a
suitable value for vmcnt().

Co-authored-by: Austin Kerbow <austin.kerbow@amd.com>.

arsenm commented on llvm/lib/Target/AMDGPU/SOPInstructions.td:
```diff
@@ -1621,6 +1621,13 @@ let OtherPredicates = [HasImageInsts] in {
+// Represents the point at which a wave must wait for all outstanding direct loads to LDS.
+// Typically inserted by the memory legalizer and consumed by SIInsertWaitcnts.
+
+def S_WAITCNT_LDS_DIRECT : SPseudoInstSI<(outs), (ins)> {
```
Synthetic variants of real instructions should use a lowercase suffix

ssahasra replied:
Yeah, I have a great deal of uncertainty on what this pseudo should be called. It seems the "_soft" suffix signifies a definite wait count argument, which the SIInsertWaitcnts pass is allowed to relax. Are you suggesting "S_WAITCNT_lds_direct"? Because "_direct" as a suffix doesn't contain enough information, and there is no "S_WAITCNT_LDS".

Pierre-vh replied:
`WAIT_LDS_DIRECT_PSEUDO` (following a similar pattern to `ATOMIC_FENCE` and other SOP pseudos?)

Pierre-vh commented on llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp:
```diff
@@ -1170,6 +1170,16 @@ bool SIGfx6CacheControl::insertWait(MachineBasicBlock::iterator &MI,
+  // SIInsertWaitcnts will later replace this with a vmcnt().
+  if (ST.hasVMemToLDSLoad() && isReleaseOrStronger(Order) &&
+      Scope == SIAtomicScope::WORKGROUP &&
+      any((SIAtomicAddrSpace)AddrSpace & SIAtomicAddrSpace::LDS)) {
```
```suggestion
      (AddrSpace & SIAtomicAddrSpace::LDS) != SIAtomicAddrSpace::NONE) {
```
Just to be consistent with the rest of the file

Pierre-vh commented on llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp:
```diff
@@ -2078,6 +2088,16 @@ bool SIGfx10CacheControl::insertWait(MachineBasicBlock::iterator &MI,
+  // SIInsertWaitcnts will later replace this with a vmcnt().
+  if (ST.hasVMemToLDSLoad() && isReleaseOrStronger(Order) &&
+      Scope == SIAtomicScope::WORKGROUP &&
+      any((SIAtomicAddrSpace)AddrSpace & SIAtomicAddrSpace::LDS)) {
```
same here

ssahasra committed:
review comments:

- try renaming to S_WAITCNT_lds_direct
- be consistent (even at the cost of brevity?)

Pierre-vh reviewed (approved):
Small nit otherwise this LGTM. Name is ok for me, I'd prefer if there was a "_pseudo" at the end instead (maybe `S_WAITCNT_LDS_DIRECT_pseudo`, just to make it clear it isn't a HW inst in any ISA, but it's really bikeshedding at this point

ssahasra committed:
remove unnecessary typecast
